BraintreeCore/src/test/java/com/braintreepayments/api/ShopperInsightsClientUnitTest.kt

```diff
@@ -0,0 +1,138 @@
+     * Tests if the getRecommendedPaymentMethods method passes correct phone body JSON to payment api
+     * when providing a phone request object.
+     */
+    @Test
+    fun testGetRecommendedPaymentMethods_verifyPhoneJson() {
+        val testCountryCode = "1"
+        val testNationalNumber = "123456789"
+        val request = ShopperInsightsRequest.Phone(
+            BuyerPhone(
+                countryCode = testCountryCode,
+                nationalNumber = testNationalNumber
+            )
+        )
+        val slot = CapturingSlot<String>()
+        sut.getRecommendedPaymentMethods(request) {
+            verify {
+                paymentApi.processRequest(capture(slot))
+            }
+        }
+
+        val customer = JSONObject(slot.captured).getJSONObject(KEY_CUSTOMER)
+        val phone = customer.getJSONObject(KEY_PHONE)
+        val countryCode = phone.getString(KEY_COUNTRY_CODE)
+        val nationalNumber = phone.getString(KEY_NATIONAL_NUMBER)
+
+        assertEquals(testCountryCode, countryCode)
+        assertEquals(testNationalNumber, nationalNumber)
+    }
+
+    /**
+     * Tests if the getRecommendedPaymentMethods method passes correct email body JSON to payment api
+     * when providing a email request object.
+     */
+    @Test
+    fun testGetRecommendedPaymentMethods_verifyEmailJson() {
+        val fakeEmail = "fake-email"
+        val request = ShopperInsightsRequest.Email(fakeEmail)
+        val slot = CapturingSlot<String>()
+        sut.getRecommendedPaymentMethods(request) {
+            verify {
+                paymentApi.processRequest(capture(slot))
+            }
+        }
+
+        val customer = JSONObject(slot.captured).getJSONObject(KEY_CUSTOMER)
+        val email = customer.getString(KEY_EMAIL)
+
+        assertEquals(fakeEmail, email)
+    }
+
+    /**
+     * Tests if the getRecommendedPaymentMethods method passes correct email and phone body JSON to payment api
+     * when providing a email and phone request object.
+     */
+    @Test
+    fun testGetRecommendedPaymentMethods_verifyEmailAndPhoneJson() {
+        val fakeEmail = "fake-email"
+        val testCountryCode = "1"
+        val testNationalNumber = "123456789"
+
+        val request = ShopperInsightsRequest.EmailAndPhone(
+            email = fakeEmail,
+            phone = BuyerPhone(
+                countryCode = testCountryCode,
+                nationalNumber = testNationalNumber
+            )
+        )
+
+        val slot = CapturingSlot<String>()
+        sut.getRecommendedPaymentMethods(request) {
+            verify {
+                paymentApi.processRequest(capture(slot))
+            }
+        }
+
+        val customer = JSONObject(slot.captured).getJSONObject(KEY_CUSTOMER)
+        val email = customer.getString(KEY_EMAIL)
+        val phone = customer.getJSONObject(KEY_PHONE)
+        val countryCode = phone.getString(KEY_COUNTRY_CODE)
+        val nationalNumber = phone.getString(KEY_NATIONAL_NUMBER)
```

Comment:
This test started out that way.
The json was initially a single line json and `detekt` didn't like it exceeded 120 characters. I had to format the json into multi-line string, that resulted in the test failing because the indentation wasn't correct. After that, I moved to using this.

Testing string equivalency for json is too strict IMO. As long as the key value pairs are correct, the json will result in the same output, so strings are a bit limited.
Now, if we had a deterministic way of converting object to json (eg: gson, moshi etc.) we could then rely on that for converting our object to strings.

Let me know if you still don't agree, and I'll try to find a better way to figure out the object to string conversion.

Reply:
Using gson / moshi would be nice, but we've always had a rule to only use support library (i.e. JetPack) dependencies in our SDK.

In our demo apps we typically use gson / moshi style libraries when communicated with our sample merchant servers, but we take a different approach in the SDK. Historically we've always done this, and it's benefited the SDK and merchants by reducing potential conflicts.

Reply:
Moved the conversion to API class.